### ISSUE TITLE
Use of Task.cpu_reqs in EnTK

### ISSUE THREAD

**Issue by ggg121** [radical-collaboration/facts#26]
**TL;DR** - How does setting "processes" to anything greater than one in `Task.cpu_reqs` affect available memory for the task in EnTK?

This may be a dumb question (and please don’t hesitate to point that out if it’s the case), but I’m a little confused by `Task.cpu_reqs` and how it relates to the resources provided to the application manager through the resource description `AppManager.resource_desc`.

My initial impression was that defining the number of CPUs in the application manager sets the total number of CPUs available for the entire workflow.  `Task.cpu_reqs` would set the number of CPUs required for the particular task.  My confusion stems from running my test_xarray workflow in EnTK on Rutger's Amarel resources.  It apparently doesn’t matter how many CPUs I define in the application manager (as long as they’re available on the remote machine), the code runs fine.  I then try to assign multiple CPUs to each task in the workflow through `Task.cpu_reqs`:

`Task.cpu_reqs = {"processes": 4, "process_type": None, "threads_per_process": 1, "thread_type": None}`

…the tasks fail to run due to “Insufficient Resources”.  I make sure that the product of “processes”, “threads_per_process”, and number of tasks is less than or equal to the number of CPUs provided to the application manager, so the only other resource I can think of that could be insufficient is memory.  If `Task.cpu_reqs` behaves how I think it does, it shouldn’t impact the amount of memory available to the particular task…which means my intuition about `Task.cpu_reqs` is incorrect.

One potential reason for the insufficient resource error I can think of would be if setting “processes” in `Task.cpu_reqs` spawns some sort of distributed memory situation (similar to MPI) where (in the example above) four instances of the task are run concurrently each with its own memory allocation.  If that’s the case, then I can definitely see where the insufficient resources error would be coming from.

Sorry for the rambling, but any insight you all can provide into this would be greatly appreciated!

**Comment by andre-merzky** [comment]
> `Task.cpu_reqs = {"processes": 4, "process_type": None, "threads_per_process": 1, "thread_type": None}`
> …the tasks fail to run due to “Insufficient Resources”.

The issue is not memory shortage - unless you specify memory requirements, RCT will schedule the tasks independent of available memory.  It seems more likely that the scheduler thinks that the nodes do not have a sufficient number of cores.  Which is strange as the requested number (4) is fairly small.  How large are the nodes?  If the task needs to be spread over multiple nodes (like, when the number of cores per node is 2), then you will need to set the `process_type` to `MPI`.

FWIW, the notation `processes` is a bit misleading - we recently renamed this attribute to `rank` as it really indicates how many MPI ranks are to be created.
